core/src/main/java/hudson/model/Run.java

```diff
@@ -1920,14 +1923,47 @@ protected void onEndBuilding() {
      */
     public synchronized void save() throws IOException {
         if(BulkChange.contains(this))   return;
-        getDataFile().write(this);
+        synchronized (saving) {
+            saving.add(this);
+        }
+        try {
+            getDataFile().write(this);
+        } finally {
+            synchronized (saving) {
+                saving.remove(this);
+            }
+        }
         SaveableListener.fireOnChange(this, getDataFile());
     }
 
     private @Nonnull XmlFile getDataFile() {
         return new XmlFile(XSTREAM,new File(getRootDir(),"build.xml"));
     }
 
+    private static final Set<Run<?, ?>> saving = new HashSet<>();
+
+    private Object writeReplace() {
+        synchronized (saving) {
+            if (saving.contains(this)) {
+                return this;
+            } else {
+                LOGGER.warning("JENKINS-45892: improper backreference detected in " + getDataFile());
```

Comment:
Note that in the common case that you just have a `[Run]Action` storing a (non`transient`) `Run owner`, `writeReplace` will get called only once—at top level—and we proceed normally. XStream then writes a `reference="../../.."` or similar, and deserialization works. We would rather print a warning in this case, but I could find no way to detect this case: [`AbstractReferenceMarshaller.convert` runs](https://github.com/jenkinsci/xstream/blob/dc387c0036c096be449c22cda833df4837097902/xstream/src/java/com/thoughtworks/xstream/core/AbstractReferenceMarshaller.java#L56-L59) and does not offer us any chance to intercept the process. This patch therefore only changes behavior in other cases:

* A build stored via some reference chain in another build, as in the test.
* A build stored in some unrelated object.
* A build stored in another copy of itself.

The third is the problematic case for lazy-loading. I am not sure of the exact sequence of events (which is why I have no test for it), but somehow a build gets loaded, a badly written action is attached to it, the build is GC’d, a new `Run` object is reloaded from disk, but we have the original `Action` somewhere, and `Run.save` is called on one copy while another copy is referred to internally. Then instead of `reference`, XStream writes out a new nested `<build>` element. When deserialized, this explodes because `onLoad` is not called on the nested `Run` and so `project` is left null…as seen in in the `toString` patch.